lib/vdsm/storage/merge.py

```diff
@@ -266,7 +266,8 @@ def finalize(subchain):
                 # optimal_size must be called when the volume is prepared
```

Comment:
We need to check if a volume can be reduced also in line 265. If a volume cannot be reduced, there is no need to get the optimal size.

Reply:
OK.

Why is the optimal size obtained within the context and the shrink called outside the context?

Reply:
To compute optimal size we measure the image using "qemu-img measure" - so we must prepare the image (activate the logical volume in block storage). Reducing the image is possible when the logical volume is active, but LVM issues warnings about this, so we prefer to teardown the volume before reducing.

The flow is:
- activate the lv and its parent lv (recursively)
- measure the qocw2 image in the lv
- deactivate the lv and its parent (recursively)
- if the lv is bigger than the optimal size, reduce it.

I guess the code needs a comment explaining this.

Reply:
Great, thanks for the explanation :)
I assumed something like that from the code, but a clear explanation is much better! I may add a comment to explain that in the code.